vim-patch:8.0.{709,728},8.1.{135,308}

janlazo opened the pull request:
**vim-patch:8.0.0709: libvterm cannot use vsnprintf()**

Problem:    Libvterm cannot use vsnprintf(), it does not exist in C90.
Solution:   Use vim_vsnprintf() instead.
https://github.com/vim/vim/commit/8327d1df1754b33d8a93b3411f30692f0042f4ce

**vim-patch:8.0.0728: the terminal structure is never freed**
Problem:    The terminal structure is never freed.
Solution:   Free the structure and unreference what it contains.
vim/vim@96ca27a

**vim-patch:8.1.0135: undo message delays screen update for CTRL-O u**
Problem:    Undo message delays screen update for CTRL-O u.
Solution:   Add smsg_attr_keep(). (closes vim/vim#3125)
vim/vim@e042968

**vim-patch:8.1.0308: a quick undo shows "1 seconds ago"**
Problem:    A quick undo shows "1 seconds ago". (Tony Mechelynck)
Solution:   Add singular/plural message.
vim/vim@fd6100b

janlazo committed:
vim-patch:8.0.0709: libvterm cannot use vsnprintf()

Problem:    Libvterm cannot use vsnprintf(), it does not exist in C90.
Solution:   Use vim_vsnprintf() instead.
https://github.com/vim/vim/commit/8327d1df1754b33d8a93b3411f30692f0042f4ce

mhinz commented:
IMHO, this patch is N/A, since we use C99 which knows `vsnprintf()`. No need to merge unneeded workarounds. :-) Especially because it's a small patch and the chance that a lot of upcoming patches rely on these changes is small.

janlazo committed:
vim-patch:8.1.0135: undo message delays screen update for CTRL-O u

Problem:    Undo message delays screen update for CTRL-O u.
Solution:   Add smsg_attr_keep(). (closes vim/vim#3125)
https://github.com/vim/vim/commit/e0429681aed5448d1c3b8399d152a2b1a05ea6d7

janlazo commented on test/functional/ui/sign_spec.lua:
```diff
@@ -239,7 +239,7 @@ describe('Signs', function()
         :sign place                                          |
         {9:--- Signs ---}                                        |
-        {10:Signs for [NULL]:}                                    |
+        {10:Signs for (null):}                                    |
```
Why did this happen?

janlazo replied:
Nevermind. Other tests failed because `vim_vsnprintf` and `vim_vsnprintf_typval` behave differently from `vsnprintf`.

bfredl replied:
An alternative is to print the buffer number if the buffer doesn't have a name. I don't think it was intended to print the internal representation of a null pointer to start with...

janlazo commented:
Tests fail if I replace `vim_vsnprintf` with `vsnprintf`.

janlazo committed:
vim-patch:8.1.0308: a quick undo shows "1 seconds ago"

Problem:    A quick undo shows "1 seconds ago". (Tony Mechelynck)
Solution:   Add singular/plural message.
https://github.com/vim/vim/commit/fd6100b2aa6178b88cfadcdbc494966bf79a5488

justinmk commented:
> Tests fail if I replace `vim_vsnprintf` with `vsnprintf`.

I don't think we need to do that, do we?

OTOH, is there any need for the extra `vim_vsnprintf_typval` function or can we just have only `vim_vsnprintf` ?

janlazo commented:
Depends on the difficulty on porting 8.1.0743 which uses `vim_vsnprintf` from this patch.

justinmk commented:
> Depends on the difficulty on porting 8.1.0743 which uses `vim_vsnprintf` from this patch.

It only saves one NULL parameter, or am I missing something?

Anyways LGTM. Is this RDY?

janlazo committed:
oldtests: set shellslash for ":cd" test

janlazo committed:
oldtests: win: filename cannot have "

janlazo commented on src/nvim/undo.c:
```diff
@@ -2587,9 +2587,13 @@ static void u_add_time(char_u *buf, size_t buflen, time_t tt)
+    int64_t seconds = time(NULL) - tt;
+    vim_snprintf((char *)buf, buflen,
+                 NGETTEXT("%" PRId64 " second ago",
+                          "%" PRId64 " seconds ago", (uint32_t)seconds),
```
gcc had issues with `uint64_t`. Does this need an assertion?

justinmk replied:
alignment issue? Maybe cast it to `int32_t` first.

janlazo committed:
vim-patch:8.0.0728: the terminal structure is never freed

Problem:    The terminal structure is never freed.
Solution:   Free the structure and unreference what it contains.
https://github.com/vim/vim/commit/96ca27a0ee8ae738cab9fb386984c75c6821e31a
